Sources/Swiftfall/Swiftfall.swift

```diff
@@ -1,635 +1,637 @@
-        let call = "cards/named?exact=\(encodeExactly)"
-        
-        var card: Card?
-        var stop = false
-        parseCard(call: call){
-            (newcard:Card?) in
-            card = newcard
-            stop = true
-        }
-        
-        while(!stop){
-            //Do this until parseCard is done
-        }
-        
-        return card
-    }
-    
-    // fuzzy
-    public static func getRandomCard() -> Card?
-    {
-        let call = "cards/random"
-        
-        var card: Card?
-        var stop = false
-        parseCard(call: call){
-            (newcard:Card?) in
-            card = newcard
-            stop = true
-        }
-        
-        while(!stop){
-            // Do this until parseCard is done
-        }
-        
-        return card
-    }
-    
-    /// Retreives JSON data from URL and parses it with JSON decoder. Thanks Mitchell
-    static func parseCard(call:String, completion: @escaping (Card?) -> ()) {
-        
-        let url = URL(string: "\(scryfall)\(call)")
-        let task = URLSession.shared.dataTask(with: url!) {(data, response, error) in
-            
-            guard let content = data else {
-                print("Error: There was no data returned from JSON file.")
-                return
-            }
-    
-            //print("\(String(data: content,encoding: .utf8))")
-            
-            let decoder = JSONDecoder()
-            do {
-                // Decode JSON file starting from Response struct.
-                let decoded:Card = try decoder.decode(Card.self, from: content)
-                completion(decoded)
-            }
-            catch {
-                // Known Issues:
-                //  * Too broad of a request (needs handler) 
-                //
-                // Present an alert if the JSON data cannot be decoded.
-                do {
-                    let decoded:Error = try decoder.decode(Error.self, from: content)
-                    decoded.simplePrint()
-                }
-                catch {
-                    print("Error: \(error)")
-                    completion(nil)
-                }
-                completion(nil)
-            }
-        }
-        task.resume()
-    }
-    
-    // set
-    public static func getSet(code: String) -> Set?
-    {
-        let encodeExactly = code.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)!
-        let call = "sets/\(encodeExactly)"
-        
-        var set: Set?
-        var stop = false
-        parseCard(call: call){
-            (newset:Set?) in
-            set = newset
-            stop = true
-        }
-        
-        while(!stop){
-            //Do this until parseCard is done
-        }
-        
-        return set
-    }
-    
-    /// Retreives JSON data from URL and parses it with JSON decoder. Thanks Mitchell
-    static func parseCard(call:String, completion: @escaping (Set?) -> ()) {
```

Comment:
This code is almost entirely the same as `parseCard(call:completion:)` above. I'd _strongly_ recommend abstracting this logic away, probably making it generic. That way the compiler can infer which type you're trying to decode based on the closure you pass in:

```swift
getResource(call: "cards/random") { (newCard: Card?) in
  // save and signal semaphore
}
```

The compiler will be able to infer based on `(newCard: Card?)` that you want it to parse a card.

```swift
static func getResource<T: Decodable>(call: String, completion: @escaping (T?) -> ()) {
  // do all the setup, then change 
  // let decoded:Set = try decoder.decode(Set.self, from: content)
  // to
  // let decoded = try decoder.decode(T.self, from: content)
}
```